node/src/coordinator.rs

```diff
@@ -0,0 +1,409 @@
+                    // For the running state, we run the full MPC protocol.
+                    // There's no timeout. The only time we stop is when the contract state
+                    // changes to no longer be running (or if somehow the epoch changes).
+                    MpcJob {
+                        name: "Running",
+                        fut: Self::create_runtime_and_run(
+                            "Running",
+                            self.config_file.cores,
+                            Self::run_mpc(
+                                self.clock.clone(),
+                                self.secret_db.clone(),
+                                self.secrets.clone(),
+                                self.config_file.clone(),
+                                self.keyshare_storage_factory.create().await?,
+                                state.clone(),
+                                self.indexer.txn_sender.clone(),
+                                self.indexer
+                                    .sign_request_receiver
+                                    .clone()
+                                    .lock_owned()
+                                    .await,
+                            ),
+                        )?,
+                        stop_fn: Box::new(move |new_state| match new_state {
+                            ContractState::Running(new_state) => new_state.epoch != state.epoch,
+                            _ => true,
+                        }),
+                        timeout_fut: futures::future::pending().boxed(),
+                    }
+                }
+                ContractState::Resharing(state) => {
+                    // In resharing state, we perform key resharing, again with a timeout.
+                    MpcJob {
+                        name: "Resharing",
+                        fut: Self::create_runtime_and_run(
+                            "Resharing",
+                            self.config_file.cores,
+                            Self::run_key_resharing(
+                                self.clock.clone(),
+                                self.secret_db.clone(),
+                                self.secrets.clone(),
+                                self.config_file.clone(),
+                                self.keyshare_storage_factory.create().await?,
+                                state.clone(),
+                                self.indexer.txn_sender.clone(),
+                            ),
+                        )?,
+                        stop_fn: Box::new(move |new_state| match new_state {
+                            ContractState::Resharing(new_state) => {
+                                new_state.old_epoch != state.old_epoch
+                                    || new_state.new_participants != state.new_participants
+                            }
+                            _ => true,
+                        }),
+                        timeout_fut: sleep(
```

Comment:
What happens if there is a fork and the node switches back and force between Resharing and Running? Would that cause some potential inconsistencies? Should we consider using final state here?

Reply:
Ah, we actually already query the final state. The contract state is proactively queried (once a second) by sending a message to the `ViewClientActor` in `get_mpc_contract_state`, and we specify Final for the finality.